test/bigquery.ts

```diff
@@ -708,26 +735,207 @@ describe('BigQuery', () => {
       );
 
       assert.throws(() => {
-        BigQuery.getType_(undefined);
+        BigQuery.getTypeDescriptorFromValue_(undefined);
+      }, expectedError);
+    });
+
+    it('should throw with an empty array', () => {
+      const expectedError = new RegExp(
+        'Type must be provided for empty array.'
+      );
+
+      assert.throws(() => {
+        BigQuery.getTypeDescriptorFromValue_([]);
       }, expectedError);
     });
+
+    it('should throw with an null value', () => {
+      const expectedError = new RegExp(
+        'Type must be provided for null values.'
+      );
+
+      assert.throws(() => {
+        BigQuery.getTypeDescriptorFromValue_(null);
+      }, expectedError);
+    });
+  });
+
+  describe('getTypeDescriptorFromProvidedType_', () => {
+    const QUERY_STRING = 'SELECT * FROM [dataset.table]';
+    const NAMED_PARAMS = {
+      key: 'value',
+    };
+    const POSITIONAL_PARAMS = ['value'];
+
+    const NAMED_TYPES = {key: 'STRING'};
+
+    const POSITIONAL_TYPES = ['STRING'];
+
+    it('should return correct type for an array', () => {
+      const type = BigQuery.getTypeDescriptorFromProvidedType_(['INT64']);
+
+      assert.deepStrictEqual(type, {
+        type: 'ARRAY',
+        arrayType: {
+          type: 'INT64',
+        },
+      });
+    });
+
+    it('should return correct type for a struct', () => {
+      const type = BigQuery.getTypeDescriptorFromProvidedType_({prop: 'INT64'});
+
+      assert.deepStrictEqual(type, {
+        type: 'STRUCT',
+        structTypes: [
+          {
+            name: 'prop',
+            type: {
+              type: 'INT64',
+            },
+          },
+        ],
+      });
+    });
+
+    it('should throw for invalid provided types with named params', done => {
+      const INVALID_TYPES = {key: 'invalid'};
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Invalid type provided./);
+      done();
+    });
+
+    it('should throw if named param not present in provided types', done => {
+      const INVALID_TYPES = {other: 'string'};
+
+      assert.throws(() => {
+        bq.createQueryJob({
```

Comment:
Sorry if I was confusing with my earlier review. Essentially, the goal is to separate tested methods as much as possible, so if we change behavior in `getTypeDescriptorFromProvidedType_()` one day, we don't have to update the tests for `createQueryJob()`. So, the tests in this describe block should call `getTypeDescriptorFromProvidedType_` directly.

When it comes to testing methods that call each other, we rely on the mocked version of the called method to verify that it is called as we expect and that the value it returns is used as we intend. That way, we don't have to know anything about what the called method is doing. It has its own `describe` that will make sure it works properly.

And then when it comes to making sure the "big picture" behavior works as intended, we have our system tests that make sure the whole machine is working as we advertise.

The tests listed below are testing functionality that lives in `createQueryJob()`, so they should be moved over there to that describe:

- should throw for invalid type structure provided for named params
- should throw for invalid type structure provided for positional params
- should throw for incorrect number of types provided for positional params

Reply:
Okay, I see.